### PR TITLE
Unit-tests for SubscriptionService, BridgeService, and TwinController

### PR DESCRIPTION
**Pull request by peolivei2**
Final round of unit-tests.

### CODE REVIEW

File: DeviceBridgeTests/Controllers/TwinControllerTests.cs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+            _twinController = new TwinController(LogManager.GetCurrentClassLogger(), _subscriptionServiceMock.Object, _bridgeServiceMock.Object);
+
+            var deviceSubscription = new DeviceSubscription();
+            _deviceSubscriptionWithStatus = new DeviceSubscriptionWithStatus(deviceSubscription);
+            _subscriptionServiceMock.Setup(s => s.GetDataSubscription(It.IsAny<Logger>(), MockDeviceId, DeviceSubscriptionType.DesiredProperties, It.IsAny<CancellationToken>())).Returns(Task.FromResult(_deviceSubscriptionWithStatus));
+            _subscriptionServiceMock.Setup(s => s.CreateOrUpdateDataSubscription(It.IsAny<Logger>(), MockDeviceId, DeviceSubscriptionType.DesiredProperties, PropertyCallbackUrl, It.IsAny<CancellationToken>())).Returns(Task.FromResult(_deviceSubscriptionWithStatus));
+        }
+
+        [Test]
+        public async Task TestGetTwin()
```
Comment (yang-hai-feng) [inline DeviceBridgeTests/Controllers/TwinControllerTests.cs:40]: Feel our device registration should only by reported/send messages, I found "get twin" will also trigger device registration, any idea?

File: DeviceBridgeTests/Controllers/TwinControllerTests.cs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+            _bridgeServiceMock = new Mock<IBridgeService>();
+            _twinController = new TwinController(LogManager.GetCurrentClassLogger(), _subscriptionServiceMock.Object, _bridgeServiceMock.Object);
+
+            var deviceSubscription = new DeviceSubscription();
+            _deviceSubscriptionWithStatus = new DeviceSubscriptionWithStatus(deviceSubscription);
+            _subscriptionServiceMock.Setup(s => s.GetDataSubscription(It.IsAny<Logger>(), MockDeviceId, DeviceSubscriptionType.DesiredProperties, It.IsAny<CancellationToken>())).Returns(Task.FromResult(_deviceSubscriptionWithStatus));
+            _subscriptionServiceMock.Setup(s => s.CreateOrUpdateDataSubscription(It.IsAny<Logger>(), MockDeviceId, DeviceSubscriptionType.DesiredProperties, PropertyCallbackUrl, It.IsAny<CancellationToken>())).Returns(Task.FromResult(_deviceSubscriptionWithStatus));
+        }
+
+        [Test]
```
Comment (yang-hai-feng) [inline DeviceBridgeTests/Controllers/TwinControllerTests.cs:39]: Add Description

File: DeviceBridgeTests/Controllers/TwinControllerTests.cs [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        {
+            _bridgeServiceMock.Setup(p => p.GetTwin(It.IsAny<Logger>(), MockDeviceId, It.IsAny<CancellationToken>())).Returns(Task.FromResult(new Twin(new TwinProperties()
+            {
+                Desired = new TwinCollection("{\"temperature\":4}"),
+            })));
+            var twinResult = await _twinController.GetTwin(MockDeviceId);
+            Assert.AreEqual("{\"twin\":{\"deviceId\":null,\"etag\":null,\"version\":null,\"properties\":{\"desired\":{\"temperature\":4},\"reported\":{}}}}", ((ContentResult)twinResult.Result).Content);
+        }
+
+        [Test]
```
Comment (yang-hai-feng) [inline DeviceBridgeTests/Controllers/TwinControllerTests.cs:50]: Add Description

File: DeviceBridgeTests/Services/BridgeServiceTests.cs [hunk 4]
```diff
@@ -1,22 +1,61 @@
...
     public class BridgeServiceTests
     {
+        private Mock<IConnectionManager> _connectionManagerMock = new Mock<IConnectionManager>();
+
+        [SetUp]
+        public async Task Setup()
+        {
+        }
+
         [Test]
```
Comment (yang-hai-feng) [inline DeviceBridgeTests/Services/BridgeServiceTests.cs:24]: Add description

File: DeviceBridgeTests/Services/BridgeServiceTests.cs [hunk 5]
```diff
@@ -1,22 +1,61 @@
...
+            var bridgeService = new BridgeService(_connectionManagerMock.Object);
+            var testPayload = new Dictionary<string, object>() { { "tel", 1 } };
+            var testDate = DateTime.UtcNow;
+            Dictionary<string, string> testProps = new Dictionary<string, string>() { { "prop", "val" } };
+            await bridgeService.SendTelemetry(LogManager.GetCurrentClassLogger(), "test-device", testPayload, default, testProps, "test-component", testDate);
+            _connectionManagerMock.Verify(p => p.AssertDeviceConnectionOpenAsync("test-device", true, false, It.IsAny<CancellationToken>()), Times.Once);
+            _connectionManagerMock.Verify(p => p.SendEventAsync(It.IsAny<Logger>(), "test-device", testPayload, default, testProps, "test-component", testDate), Times.Once);
+        }
+
+        [Test]
```
Comment (yang-hai-feng) [inline DeviceBridgeTests/Services/BridgeServiceTests.cs:37]: Add Description

File: DeviceBridgeTests/Services/BridgeServiceTests.cs [hunk 6]
```diff
@@ -1,22 +1,61 @@
...
+            var bridgeService = new BridgeService(_connectionManagerMock.Object);
+            var testTwin = new Twin();
+            _connectionManagerMock.Setup(p => p.GetTwinAsync(It.IsAny<Logger>(), "test-device", It.IsAny<CancellationToken>())).Returns(Task.FromResult(testTwin));
+            var result = await bridgeService.GetTwin(LogManager.GetCurrentClassLogger(), "test-device", default);
+            Assert.AreEqual(testTwin, result);
+            _connectionManagerMock.Verify(p => p.AssertDeviceConnectionOpenAsync("test-device", true, false, It.IsAny<CancellationToken>()), Times.Once);
+            _connectionManagerMock.Verify(p => p.GetTwinAsync(It.IsAny<Logger>(), "test-device", default), Times.Once);
+        }
+
+        [Test]
```
Comment (yang-hai-feng) [inline DeviceBridgeTests/Services/BridgeServiceTests.cs:50]: Add Description

File: DeviceBridgeTests/Services/BridgeServiceTests.cs [hunk 7]
```diff
@@ -1,22 +1,64 @@
...
 namespace DeviceBridge.Services.Tests
 {
     [TestFixture]
     public class BridgeServiceTests
     {
+        private Mock<IConnectionManager> _connectionManagerMock = new Mock<IConnectionManager>();
+
+        [SetUp]
+        public async Task Setup()
+        {
```
Comment (Jackbk-nz) [inline DeviceBridgeTests/Services/BridgeServiceTests.cs:21]: Can remove empty setup method

File: DeviceBridgeTests/Services/SubscriptionServiceTests.cs [hunk 8]
```diff
@@ -0,0 +1,379 @@
...
+{
+    [TestFixture]
+    public class SubscriptionServiceTests
+    {
+        private Mock<IStorageProvider> _storageProviderMock = new Mock<IStorageProvider>();
+        private Mock<IHttpClientFactory> _httpClientFactoryMock = new Mock<IHttpClientFactory>();
+        private Mock<IConnectionManager> _connectionManagerMock = new Mock<IConnectionManager>();
+
+        [SetUp]
+        public async Task Setup()
```
Comment (Jackbk-nz) [inline DeviceBridgeTests/Services/SubscriptionServiceTests.cs:31]: +1

**Review by Jackbk-nz — APPROVED** [review]

File: DeviceBridgeTests/Controllers/TwinControllerTests.cs (hunk 1 above)
Comment (peolivei2) [reply DeviceBridgeTests/Controllers/TwinControllerTests.cs:41]: Synced offline. Added a note to the README in the API section about automatic registration.

**Pull request merged**